File: streamflowSignatures/julia/src/timing.jl
"""
Flow timing signatures.

Cumulative flow timing metrics (D-day metrics) and peak timing.
"""

"""
    analyze_flow_timing_trends(df::DataFrame) -> Dict

Calculate flow timing signatures with trend statistics.

Calculates 15 metrics:
- D1_day through D99_day: Days when cumulative flow reaches X% (13 D-day
  percentiles from config `timing.d_percentiles`)
- D25_to_D75: Duration of middle 50% of flow
- Dmax: Day of maximum flow

Each metric produces 8 statistics via generate_stats().

Parameters
----------
df : DataFrame
    Daily streamflow data with columns: water_year, Q, dowy
    (preprocessor guarantees all years are valid)

Returns
-------
Dict{String, Float64}
    Dictionary of signature statistics (15 metrics × 8 stats = 120 values)
"""
function analyze_flow_timing_trends(df::DataFrame; trend_completeness::Union{Nothing, Float64}=nothing, decade_completeness::Union{Nothing, Float64}=nothing, min_values_for_stats::Union{Nothing, Int}=nothing, changepoint::Union{Nothing, NamedTuple}=nothing, collector::Union{Nothing, AnnualCollector}=nothing)
    result = Dict{String, Float64}()

    # D-day percentiles from config
    d_percentiles = CFG_D_PERCENTILES
    metrics = ["D$(p)_day" for p in d_percentiles]
    push!(metrics, "D25_to_D75", "Dmax")

    # Column validation
    valid, missing_cols = validate_columns(df, ["Q", "water_year", "dowy"])
    if !valid
        @warn "analyze_flow_timing_trends: Missing columns: $missing_cols"
        for m in metrics
            merge!(result, empty_stats(m))
        end
        return result
    end

    if nrow(df) == 0
        for m in metrics
            merge!(result, empty_stats(m))
        end
        return result
    end

    years = unique(df.water_year)
    # Build DataFrame dynamically from config percentiles
    annual_data = DataFrame(water_year = years)
    for p in d_percentiles
        annual_data[!, Symbol("D$(p)_day")] = fill(NaN, length(years))
    end
    annual_data[!, :D25_to_D75] = fill(NaN, length(years))
    annual_data[!, :Dmax] = fill(NaN, length(years))

    for yr in years
        year_mask = coalesce.(df.water_year .== yr, false)
        if !any(year_mask)
            continue
        end
        year_df = df[year_mask, :]

        yr_idx = findfirst(annual_data.water_year .== yr)
        yr_idx === nothing && continue

        # Sort by day of water year (handle Missing dowy)
        dowy_clean = [coalesce(d, 999) for d in year_df.dowy]
        perm = sortperm(dowy_clean)
        dowy_sorted = dowy_clean[perm]

        # Handle NaN/Missing in Q - skip year if any NaNs remain
        # (preprocess_daily_data guarantees zero NAs for valid years)
        Q_values = coalesce_q(year_df.Q)[perm]
        n_na = count(isnan, Q_values)
        if n_na > 0
            continue
        end

        # Cumulative flow
        cum_Q = cumsum(Q_values)
        total_Q = cum_Q[end]

        if total_Q <= 0
            continue
        end

        # Normalize to percentage
        cum_pct = cum_Q ./ total_Q .* 100

        # Find D-day for each percentile
        for p in d_percentiles
            # Find first day where cumulative flow >= p%
            idx = findfirst(cum_pct .>= p)
            if idx !== nothing
                annual_data[yr_idx, Symbol("D$(p)_day")] = dowy_sorted[idx]
            end
        end

        # D25 to D75 duration — compute directly from cumulative percentages
        # (NOT from D_PERCENTILES, which may not include 25 and 75)
        idx_25 = findfirst(cum_pct .>= 25)
        idx_75 = findfirst(cum_pct .>= 75)
        if idx_25 !== nothing && idx_75 !== nothing
            annual_data[yr_idx, :D25_to_D75] = dowy_sorted[idx_75] - dowy_sorted[idx_25]
        end

        # Day of maximum flow
        max_idx = argmax(Q_values)
        annual_data[yr_idx, :Dmax] = dowy_sorted[max_idx]
    end

    if nrow(annual_data) < 3
        for m in metrics
            merge!(result, empty_stats(m))
        end
        return result
    end

    # Generate statistics for all metrics
    value_cols = filter(x -> x != "water_year", names(annual_data))
    result = generate_stats(annual_data; value_cols=value_cols, trend_completeness=trend_completeness, decade_completeness=decade_completeness, min_values_for_stats=min_values_for_stats, changepoint=changepoint, collector=collector)

    return result
end

File: streamflowSignatures/python/streamflow_signatures/timing.py
"""
Flow timing signatures.

Calculates when during the water year cumulative flow reaches
various percentile thresholds (center of mass timing).
"""

from typing import Dict, Optional
import numpy as np
import pandas as pd
from .stats import generate_stats
from .config import D_PERCENTILES


def analyze_flow_timing_trends(
    streamflow_data: pd.DataFrame,
    trend_completeness: Optional[float] = None,
    decade_completeness: Optional[float] = None,
    min_values_for_stats: Optional[int] = None,
    changepoint: Optional[dict] = None,
    collector: Optional[object] = None,
) -> Dict[str, float]:
    """
    Calculate flow timing metrics trends.

    Determines the day of water year when cumulative flow reaches
    various percentile thresholds (1%, 5%, 10%, ..., 95%, 99%) —
    15 metrics: 13 D-day percentiles (config timing.d_percentiles)
    plus D25_to_D75 and Dmax.

    Parameters
    ----------
    streamflow_data : pd.DataFrame
        Daily streamflow data with columns:
            - water_year: int, water year
            - Q: float, daily discharge in mm/day
            - dowy: int, day of water year (1-366)

    Returns
    -------
    dict
        Dictionary of signature statistics with keys for each metric:
            - D1_day, D5_day, ..., D99_day: Day when cumulative
              flow reaches each percentile
            - D25_to_D75: Duration between 25% and 75% cumulative flow
            - Dmax: Day of maximum discharge

    Notes
    -----
    - Day 1 = October 1 (start of water year)
    - Day 365/366 = September 30 (end of water year)
    - Years with any remaining NAs after coercion are skipped
    """
    # Validate required columns
    required_cols = ["water_year", "Q", "dowy"]
    missing = [c for c in required_cols if c not in streamflow_data.columns]
    if missing:
        raise ValueError(f"Missing required columns: {missing}")

    df = streamflow_data

    # Get unique years
    years = df["water_year"].unique()

    # Define percentiles to calculate (from config)
    percentiles = D_PERCENTILES

    # Initialize results DataFrame
    columns = ["water_year"] + [f"D{p}_day" for p in percentiles] + ["D25_to_D75", "Dmax"]
    timing_by_year = pd.DataFrame(index=range(len(years)), columns=columns)
    timing_by_year["water_year"] = years

    # Build a lookup from water_year to row index for fast assignment
    yr_to_idx = {yr: i for i, yr in enumerate(years)}

    # Calculate timing metrics for each year
    for yr, year_data in df.groupby("water_year", sort=False):

        idx = yr_to_idx[yr]

        # Sort by day of water year
        year_data = year_data.sort_values("dowy")

        # Skip years with any NAs in Q
        q_vals = year_data["Q"].values
        if np.any(np.isnan(q_vals)):
            continue

        # Calculate total annual flow
        total_flow = np.sum(q_vals)

        # Skip years with zero or negative total flow
        if total_flow <= 0:
            continue

        cum_flow = np.cumsum(q_vals)

        # Calculate cumulative flow as percentage of total
        cum_pct = (cum_flow / total_flow) * 100

        dowy_values = year_data["dowy"].values

        # Find day when cumulative flow reaches each percentile
        for p in percentiles:
            above_threshold = np.where(cum_pct >= p)[0]
            if len(above_threshold) > 0:
                timing_by_year.loc[idx, f"D{p}_day"] = dowy_values[above_threshold[0]]

        # Calculate D25_to_D75 (days between 25% and 75% cumulative flow)
        above_25 = np.where(cum_pct >= 25)[0]
        above_75 = np.where(cum_pct >= 75)[0]

        if len(above_25) > 0 and len(above_75) > 0:
            day_25 = dowy_values[above_25[0]]
            day_75 = dowy_values[above_75[0]]
            timing_by_year.loc[idx, "D25_to_D75"] = day_75 - day_25

        # Calculate Dmax (day of maximum discharge)
        q_values = year_data["Q"].values
        if not np.all(np.isnan(q_values)):
            max_idx = np.nanargmax(q_values)
            timing_by_year.loc[idx, "Dmax"] = dowy_values[max_idx]

    # Convert columns to numeric
    metric_cols = [f"D{p}_day" for p in percentiles] + ["D25_to_D75", "Dmax"]
    for col in metric_cols:
        timing_by_year[col] = pd.to_numeric(timing_by_year[col], errors="coerce")
    timing_by_year["water_year"] = pd.to_numeric(timing_by_year["water_year"])

    # Generate statistics
    return generate_stats(
        timing_by_year,
        value_cols=metric_cols,
        year_col="water_year",
        trend_completeness=trend_completeness,
        decade_completeness=decade_completeness, min_values_for_stats=min_values_for_stats, changepoint=changepoint, collector=collector,
    )
